src/data/fin.lean

```diff
@@ -1012,6 +1012,168 @@ def append {α : Type*} {o : ℕ} (ho : o = m + n) (u : fin m → α) (v : fin n
+    { erw function.update_same,
+      convert function.update_same i _ _,
+      ext,
+      refl },
+    { rw [function.update_noteq, function.update_noteq, append_apply_fst ho _ h],
+      all_goals { contrapose! h_1, try {rw h_1 <|> rw ←h_1}, ext, refl }}},
+  { rw [append_apply_snd ho _ h, function.update_noteq],
+    { rw append_apply_snd ho _ h },
+    { contrapose! h, exact h.symm ▸ i.2 }},
+end
+
+lemma append_update_snd (ho : o = m + n) (i : fin n) (x : ι) :
+  append ho u (function.update v i x) =
+    function.update (append ho u v) (cast ho.symm (nat_add m i)) x :=
+begin
+  ext y,
+  cases classical.em ((y : ℕ) < m),
+  { rw [append_apply_fst ho y h, function.update_noteq, append_apply_fst],
+    contrapose! h,
+    rw h,
+    exact nat.le_add_right _ _ },
+  { rw append_apply_snd,
+    cases classical.em (y = (cast ho.symm (nat_add m i))) with hi hi,
+    { have hy : (⟨(y : ℕ) - m, (nat.sub_lt_left_iff_lt_add $
+        le_of_not_lt h).2 (ho ▸ y.2)⟩ : fin n) = i :=
+      fin.ext (by rw [coe_mk, hi]; exact nat.add_sub_cancel_left _ _),
+      rw [hy, function.update_same, hi, function.update_same] },
+    { rw [function.update_noteq hi, function.update_noteq, append_apply_snd ho _ h],
+      contrapose! hi,
+      rw ←hi,
+      ext,
+      exact (nat.add_sub_of_le (not_lt.1 h)).symm }},
+end
+
+lemma append_assoc {k p r s : ℕ} {w : fin k → ι} (ho : o = m + n)
+  (hp : p = o + k) (hr : r = n + k) (hs : s = m + r) (x) :
+  append hp (append ho u v) w x = append hs u (append hr v w)
+    (cast (by rw [hp, hs, hr, ho, add_assoc]) x) :=
+begin
+  cases classical.em ((x : ℕ) < o) with hx hx,
+  { rw append_apply_fst hp _ hx,
+    cases classical.em ((x : ℕ) < m) with hmx hmx,
+    { rw [append_apply_fst hs, append_apply_fst ho],
+      congr,
+      exact hmx },
+    { rw [append_apply_snd ho, append_apply_snd hs, append_apply_fst hr],
+      congr,
+      all_goals { exact hmx }}},
+  { rw [append_apply_snd hp, append_apply_snd hs, append_apply_snd hr],
+    { congr' 1,
+      ext,
+      simp only [coe_mk, nat.sub_sub],
+      exact congr_arg _ ho },
+    { rw coe_mk,
+      contrapose! hx,
+      rw ho,
+      exact nat.lt_add_of_sub_lt_left hx },
+    { contrapose! hx,
+      rw ho,
+      exact lt_of_lt_of_le hx (le_add_right $ le_refl _) },
+    { exact hx }},
+end
+
+lemma append_assoc' {k p r s : ℕ} {w : fin k → ι} (ho : o = m + n)
+  (hp : p = o + k) (hr : r = n + k) (hs : s = m + r) (x) :
+  append hs u (append hr v w) x = append hp (append ho u v) w (cast
+    (by rw [hp, hs, hr, ho, add_assoc]) x) :=
+(append_assoc ho hp hr hs (cast (by rw [hp, hs, hr, ho, add_assoc]) x)).symm
+
+lemma append_cons_fst {p : ℕ} (ho : o = m.succ + n) (hp : p = m + n) {x : ι} {i : fin o} :
+  (append ho (cons x u : fin m.succ → ι) v) i = (cons x (append hp u v) : fin p.succ → ι)
+    (cast (by rw [ho, nat.succ_add, ←hp]) i) :=
+begin
+  cases classical.em ((i : ℕ) < m.succ),
```

Comment:
No need for choice here:
```suggestion
  cases decidable.em ((i : ℕ) < m.succ),
```
or:
```suggestion
  by_cases (i : ℕ) < m.succ,
```
which automatically chooses whether to use `classical` or `decidable` automatically.